### PR TITLE
[CMS-834] Hook into graphql to emit the right surrogate keys.

### PR DESCRIPTION
**Pull request by kporras07**

### CODE REVIEW

File: inc/class-emitter.php [hunk 1]
```diff
@@ -371,4 +378,55 @@ function( $a, $b ) {
...
+		$keys = apply_filters( 'pantheon_wp_graphql_surrogate_keys', $keys );
+		$keys = array_unique( $keys );
+		$keys = self::filter_huge_surrogate_keys_list( $keys );
+		return $keys;
+	}
+
+	/**
+	 * Send additional headers to graphql response.
+	 */
+	public static function filter_graphql_response_headers_to_send($headers) {
```
Comment (jspellman814) [inline inc/class-emitter.php:425]:
```suggestion
	public static function filter_graphql_response_headers_to_send( $headers ) {
```

File: inc/class-emitter.php [hunk 2]
```diff
@@ -371,4 +378,55 @@ function( $a, $b ) {
 
 		return $keyout;
 	}
+
+	/**
+	 * Inspect the model and get the right surrogate keys.
+	 */
+	public static function filter_graphql_dataloader_get_model( $model ) {
+		$reflect = new \ReflectionClass($model);
```
Comment (jspellman814) [inline inc/class-emitter.php:386]:
```suggestion
		$reflect = new \ReflectionClass( $model );
```

File: inc/class-emitter.php [hunk 3]
```diff
@@ -371,4 +378,55 @@ function( $a, $b ) {
...
+		$class_short_name = $reflect->getShortName();
+		$surrogate_key_prefix = strtolower($class_short_name);
+		if ( isset( $model->id )) {
+			if (!empty($model->databaseId)) {
```
Comment (jspellman814) [inline inc/class-emitter.php:390]:
```suggestion
			if ( ! empty( $model->databaseId ) ) {
```

File: inc/class-emitter.php [hunk 4]
```diff
@@ -371,4 +378,55 @@ function( $a, $b ) {
...
+				self::get_instance()->graphql_surrogate_keys[] = $surrogate_key_prefix . '-' . $model->databaseId;
+			}
+			if (!empty($model->slug)) {
```
Comment (jspellman814) [inline inc/class-emitter.php:393]:
```suggestion
			if ( ! empty( $model->slug ) ) {
```

File: inc/class-emitter.php [hunk 5]
```diff
@@ -371,4 +378,55 @@ function( $a, $b ) {
...
 		return $keyout;
 	}
+
+	/**
+	 * Inspect the model and get the right surrogate keys.
+	 */
+	public static function filter_graphql_dataloader_get_model( $model ) {
+		$reflect = new \ReflectionClass($model);
+		$class_short_name = $reflect->getShortName();
+		$surrogate_key_prefix = strtolower($class_short_name);
```
Comment (jspellman814) [inline inc/class-emitter.php:388]:
```suggestion
		$surrogate_key_prefix = strtolower( $class_short_name );
```

**Review by jspellman814 — CHANGES_REQUESTED** [review]
~~Few small formatting changes to maintain consistency with the rest of the plugin.~~ Linting picked up on these 😄  Otherwise looking good to me.

**Review by jspellman814 — APPROVED** [review]
LGTM! Let's hold on merging until we hear from Decoupled, though.

**Review by backlineint — CHANGES_REQUESTED** [review]
I see the surrogate key header on wp-graphql responses, which is great. And it seems to handle cases with multiple results well. But I'm seeing more keys than I would expect. 

Take an individual piece of page content - if I look at the header when viewing the page in WordPress, I see the following keys:

post-2 post-user-1 single

If I use a graphql query for that page:

```
query NewQuery {
  page(id: "cG9zdDoy") {
        id
        content
  }
}
```

I see the following keys:

post-2 post-sample-page user-1 user-decoupledadmin

That more than doubles the header, with some info that I don't thing Fastly would use. Not a huge deal for one page, but I wonder if it could be an issue for a query that returns something like 50 results. From the plugin docs:

> There is a limit to the number of surrogate keys in a response, so we’ve optimized them based on a user’s expectation of a normal WordPress site

Also, I wonder about leaking the username here.

So I'd be interested to know how much effort would be required to remove the keys that aren't included for the standard WP response. That seems more consistent and efficient. That said, the current implementation is really close and would likely work in the majority of cases.

**Comment by kporras07** [comment]
@backlineint Quick question:

Currently, we have:
- Page: post-2 post-user-1 single
- GraphQL: post-2 post-sample-page user-1 user-decoupledadmin

I can easily make GraphQL to look like this: post-2 user-1

1) Should I also change “user-1”  to “post-user-1"?
2) Do you really need the "single" key to be outputted via graphql? (That one will likely take some more effort)

**Comment by kporras07** [comment]
I went ahead and removed the slug keys and added post-user-* keys for posts (lmk if we need this for any other "entity type"). Also, I added some code that will remove a "user-*" key if there is a corresponding "post-user-*" key.

Currently open questions:
1) Should we add any other keys similar to "post-user-*"?
2) Should we add a "single" key under certain conditions?

**Comment by kporras07** [comment]
@backlineint updated to match recent Slack conversation.

Example output:

```
post-9 post-1 user-1 graphql-collection
```

**Comment by kporras07** [comment]
It was a request from @backlineint when I asked him via Slack about this

**Review by backlineint — APPROVED** [review]
Did some testing with the latest - looks right to me. Thanks for the quick revisions!

**Pull request merged**